albatross/outlier.h

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (C) 2018 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_CORE_OUTLIER_H
+#define ALBATROSS_CORE_OUTLIER_H
+
+#include "core/dataset.h"
+#include "core/indexing.h"
+#include "core/model.h"
+#include "crossvalidation.h"
+#include "random_utils.h"
+#include <functional>
+#include <random>
+#include <set>
+
+namespace albatross {
+
+using Indexer = std::vector<std::size_t>;
+using GroupIndexer = std::vector<std::vector<std::size_t>>;
+
+Indexer concatenate_subset_of_groups(const Indexer &subset_indices,
+                                     const GroupIndexer &indexer) {
+
+  Indexer output;
+  for (const auto i : subset_indices) {
+    assert(i < static_cast<std::size_t>(indexer.size()));
+    output.insert(output.end(), indexer[i].begin(), indexer[i].end());
+  }
+  return output;
+}
+
+/*
+ * This RANdom SAmple Consensus (RANSAC) algorithm works as follows.
+ *
+ *   1) Randomly sample a small number of data points and fit a
+ *      reference model to that data.
+ *   2) Assemble all the data points that agree with the
+ *      reference model into a set of inliers.
+ *   3) Evaluate the quality of the inliers.
+ *   4) Repeat N times keeping track of the best set of inliers.
+ *
+ * One of the large drawbacks of this approach is the computational
+ * load since it requires fitting and predicting repeatedly.
+ * The goal of this implementation is to provide a way
+ * for the user to optionally perform a lot of computation upfront,
+ * then use call backs which take indices as input to selectively
+ * update/downdate the model to produce the fits and evaluation
+ * metrics.
+ */
+template <typename FitType>
+Indexer
+ransac(std::function<FitType(const Indexer &)> &fitter,
+       std::function<double(const Indexer &, const FitType &)> &outlier_metric,
+       std::function<double(const Indexer &)> &model_metric,
+       const GroupIndexer &indexer, double threshold, std::size_t min_features,
+       std::size_t min_inliers, std::size_t max_iterations) {
+
+  std::default_random_engine gen;
+
+  Indexer reference;
+  double best_metric = HUGE_VAL;
```

Comment:
Could use an optional here?

Reply:
I haven't included `optional` in `albatross` yet.  There have been a few use cases ... but I'm not sure enough that it seems worth the added dependency.